gnc/sub8_perception/src/sub8_pcl/sub8_gateway.py

```diff
@@ -0,0 +1,202 @@
+__author__ = 'Erol Bahadiroglu'
+
+'''
+This script receives images from one of the submarine's cameras (stereo/left) and searches for a shape resembles a
+gateway. This involves applying the Harris Corner Detection tool (from OpenCV), constructing a set of vectors between
+the corner points, and finding overlapping pairs of perpendicular vectors.
+
+Output is sent with the topic 'gatway'. It is a JSON with the format [[top left corner], [dimensions]].
+'''
+
+import cv2
+import matplotlib.pyplot as plt
+import numpy as np
+import rospy
+from sensor_msgs.msg import Image as ROS_img
+from std_msgs.msg import String as ROS_string
+import cv_bridge
+import json
+
+camera = 'stereo/left'  # address of camera
+destination = 'gateway'  # topic where data is published
+#input_image = cv2.imread("images.png")  # testing image
+
+############### IMAGE PROCESSING ###############
+
+class Vector(object):
+    def __init__(self, unit_vector, magnitude, point1, point2):
+        self.unit = unit_vector
+        self.magnitude = magnitude
+        self.point_set = (point1, point2)
+
+    def __repr__(self):
+        return "unit vector: " + str(self.unit) + "; magnitude: " + str(self.magnitude) + "; " + str(self.point_set)
+
+    def equals(self, other):
+        if (self.magnitude == other.magnitude) and (set(self.unit) == set(other.unit)):
+            return True
+        else:
+            return False
+
+    @staticmethod
+    def unit_dot(v1, v2):
+        return (v1.unit[0] * v2.unit[0]) + (v1.unit[1] * v2.unit[1])
+
+    @staticmethod
+    def is_ortho(v1, v2, tolerance):
+        product = Vector.unit_dot(v1,v2)
+        if abs(product) < tolerance:
+            return True
+        else:
+            return False
+
+    @staticmethod
+    def is_parallel(v1, v2, tolerance):
+        product = Vector.unit_dot(v1,v2)
+        if (1 - abs(product)) < tolerance:
+            return True
+        else:
+            return False
+
+def color_filter(image, target_hue=None, t_hue=None, target_saturation=None, t_sat=None, target_brightness=None, t_brt=None):
+    hsv_image = cv2.cvtColor(image, cv2.COLOR_BGR2HSV)
+    masks = []
+    if target_hue:
+        hue_matrix = hsv_image[:,:,0]
+        mask = cv2.inRange(hue_matrix, target_hue-t_hue, target_hue+t_hue)

```

Comment:
inRange can be used on a numpy array,

ex:
    mask = cv2.inRange(img, np.array([0, 0, 0]), np.array([255, 255, 255]))